### PR TITLE
feat: stoplight project authorization

### PR DESCRIPTION
**Pull request by mallachari**
Fixes: https://github.com/stoplightio/elements/issues/1361

Adds `authToken` property to `StoplightProject` component and uses it with fetch handlers to platform endpoints.
Tested with private projects on running platform.

### CODE REVIEW

**Review by marcelltoth — COMMENTED** [review]
I'd like to stop for a second to think about the naming of this prop. There are some proposals (https://github.com/stoplightio/elements/issues/1400) to add API related auth functionality to `StoplightProject / API`. 

We may or may not decide to implement such, but in order to protect us from possible name collisions we could consider naming this something like `platformAuthToken` or similar, lining up with `platformUrl`. @philsturgeon WDYT?

**Review by mmiask — APPROVED** [review]
Awesome stuff! Looks good, has a test covering the expected behavior - let's merge it!

**Review by marcelltoth — CHANGES_REQUESTED** [review]
To wait for @philsturgeon 's feedback on the prop name

**Comment by philsturgeon** [comment]
I’ve got to admit I’m a bit confused about the auth work we need to do. I’m not sure I fully understand the technical requirements which has an impact on the product requirements.

I thought we wanted to look for JWTs being returned, and pass them back to the service in subsequent requests so that “logged in users” could get relevant content for them, internal, private, etc. It seems like this might not have been correct. 

The most vague suggestion I could make is that if the authToken is the only type of authToken we’ll need then there’s no need to make the variable name longer for clarification.

**Comment by philsturgeon** [comment]
Btw the api interactions will always be limited to StoplightProject. The API component should not ever need any Stoplight API interactions to run, it should always be standalone.

**Comment by mpodlasin** [comment]
> I’ve got to admit I’m a bit confused about the auth work we need to do. I’m not sure I fully understand the technical requirements which has an impact on the product requirements.
> 
> I thought we wanted to look for JWTs being returned, and pass them back to the service in subsequent requests so that “logged in users” could get relevant content for them, internal, private, etc. It seems like this might not have been correct.
> 
> The most vague suggestion I could make is that if the authToken is the only type of authToken we’ll need then there’s no need to make the variable name longer for clarification.

You are essentially correct, but we will not perform authentication (getting a token for the user in platform), because platform already does it.

Platform already has a token and it will pass it to dev-portal components.

@philsturgeon is however kind of correct here - this prop shouldn't be on the `StoplightProject` component itself, because this component is not used in platform. It should be passed only via `DevPortalProvider`, which will be used by platform when using components such as `TableOfContents`, `NodeContent` etc.

This also means that there will be no auth token naming conflict - `authToken` on `DevPortalProvider` is completely enough, but `platformAuthToken` is probably better, because it's more explicit what kind of token is that.

Second auth token that @marcelltoth mentions *will be* on StoplightProject component (and API for that matter) - given we accept that request from the user. But that's a whole separate issue.

**Review by mpodlasin — CHANGES_REQUESTED** [review]
Look at my other comment.

There is no reason for `authToken` to be a `StoplightProject` prop.

Also cc @lottamus to make sure what we are doing here makes sense.

**Comment by mallachari** [comment]
So we don't want StoplightProject to be able to access private projects for external users?

**Comment by mpodlasin** [comment]
"so Platform can pass in the user’s JWT." - this is what the issue says. To achieve that prop on the StoplightProject is unnecessary.

I guess we could allow private docs to be displayed in StoplightProject, but how to achieve that is a whole separate conversation IMHO.

**Comment by mallachari** [comment]
Well then let's have it just in provider. I assumed that people using docs might be in possession of token by doing some kind of external login. But if we don't need that now that's fine, let's focus on making it usable for platform.

**Comment by mpodlasin** [comment]
> Well then let's have it just in provider. I assumed that people using docs might be in possession of token by doing some kind of external login. But if we don't need that now that's fine, let's focus on making it usable for platform.

And note that this will have a big effect on your unit tests.

There should be more tests that hooks pass the authorization token when they are in the `DevPortalProvider` context.

**Comment by lottamus** [comment]
> There is no reason for authToken to be a StoplightProject prop.

Correct, we don't need `authToken` on `StoplightProject` (yet, see below). Platform doesn't use `StoplightProject`, instead it uses the data fetching handlers and passes data directly into TableOfContents and NodeContent components. So we just need to support `authToken` (or some way to pass the user's auth token) in the handlers, so we can send the Authorization header in the HTTP request to Elements API.

> So we don't want StoplightProject to be able to access private projects for external users?

Yeah, this is a bit trickier. We don't expose a way for folks to retrieve an auth token right now. Plus we don't want folks to expose their auth tokens to the public if providing it via a prop on the component. We'll need to think through the best approach to supporting private projects, but it will likely come with some sort of Login/Auth component where folks are required to log into their Stoplight account.

File: packages/elements-dev-portal/src/hooks/dataFetching.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  beforeEach(() => {
+    fetchMock.mockResponse(() => Promise.resolve({ status: 200 }));
+    fetchMock.enableMocks();
+  });
+
+  afterEach(() => {
+    fetchMock.disableMocks();
+  });
+
+  it.each<[string, Function, Record<string, string>]>([
```
Comment (mpodlasin) [inline packages/elements-dev-portal/src/hooks/dataFetching.spec.tsx:23]: Oh, this looks great!

**Review by mpodlasin — APPROVED** [review]
Looks great!

**Pull request merged**